src/backend/distributed/planner/query_pushdown_planning.c

```diff
@@ -1397,6 +1427,40 @@ RelationInfoContainsOnlyRecurringTuples(PlannerInfo *plannerInfo, Relids relids)
 }
 
 
+/*
+ * DeferredErrorIfUnsupportedLateralSubquery returns true if any of the
+ * relations in a RelOptInfo contain a lateral subquery that has a merge step.
+ */
+static DeferredErrorMessage *
+DeferredErrorIfUnsupportedLateralSubquery(PlannerInfo *plannerInfo, Relids relids)
+{
+	int relationId = -1;
+
+	while ((relationId = bms_next_member(relids, relationId)) >= 0)
+	{
+		RangeTblEntry *rangeTableEntry = plannerInfo->simple_rte_array[relationId];
+
+		if (!rangeTableEntry->lateral)
+		{
+			continue;
+		}
+
+		/* TODO: What about others kinds? */
```

Comment:
bool		lateral;		/* subquery, function, or values is LATERAL? */

From `RangeTblEntry`

Reply:
do we need to consider something like `LEFT JOIN LATERAL fn((SELECT count(*) FROM dist)) ON (true)`?

Reply:
As of today, we replace FUNCTIONs in the JOIN trees with subqueries automatically: https://github.com/citusdata/citus/blob/master/src/backend/distributed/planner/recursive_planning.c#L1588

But seems to ignore LATERALS: https://github.com/citusdata/citus/blob/master/src/backend/distributed/planner/recursive_planning.c#L1803-L1812

So, definitely worth testing but I wouldn't mind if it fails

Reply:
Seems this is incorrectly allowed:
```sql
 CREATE OR REPLACE FUNCTION fn(x int)
RETURNS int
AS $$
DECLARE
BEGIN
        RETURN x + 1;
END;
$$ LANGUAGE plpgsql;

SELECT count(*)
FROM ref JOIN test on ref.b = test.x,
    LATERAL fn((
        SELECT
            test_2.x
        FROM test test_2
        WHERE
            test_2.x = test.x
        LIMIT 1
    )) q
;
```

Reply:
Yes, but it's not just this new code that doesn't handle it. It doesn't even get to the `DeferredErrorIfUnsupportedLateralSubquery` check. This is because `RelationInfoContainsOnlyRecurringTuples` is not handling this correctly. I discussed with Onder and this query is also wrong, which even uses a LEFT JOIN.
```
SELECT count(*)
FROM ref LEFT JOIN GREATEST(
        0,
        (SELECT
            test.y
        FROM test
        WHERE
            test.y = ref.a
        LIMIT 1)
    ) ON TRUE;
```

Reply:
Without the left join also returns incorrect results:
```sql
SELECT count(*)
FROM ref,
    LATERAL GREATEST(
        0,
        (SELECT
            test.y
        FROM test
        WHERE
            test.y = ref.a
        LIMIT 1)
    );
```

Reply:
Another one that returns wrong results:
```sql
SELECT count(*)
FROM (select GREATEST(1, 3)) ref(a),
    LATERAL (
        SELECT
            test.y
        FROM test
        WHERE
            test.y = ref.a
        LIMIT 2
    ) q;
```

Reply:
For future reference: Our brief check with Jelte showed that there are `SubPlans` and they do not show up in `plannerInfo->simple_rte_array[relationId];` So, we might need to include the Rtes of subplans on all the places we rely on `plannerInfo->simple_rte_array[relationId];`

Reply:
some elaboration on the TODO with this context would be helpful